rust/rbac-registration/src/registration/cardano/validation.rs

```diff
@@ -0,0 +1,326 @@
+    registration::cardano::RegistrationChain,
+};
+
+/// A return value of the `validate_rbac_registration` method.
+pub type RbacValidationResult = Result<RbacValidationSuccess, RbacValidationError>;
+
+/// An error returned from the `validate_rbac_registration` method.
+#[allow(clippy::large_enum_variant)]
+pub enum RbacValidationError {
+    /// A registration is invalid (`report.is_problematic()` returns `true`).
+    ///
+    /// This variant is inserted to the `rbac_invalid_registration` table.
+    InvalidRegistration {
+        /// A Catalyst ID.
+        catalyst_id: CatalystId,
+        /// A registration purpose.
+        purpose: Option<UuidV4>,
+        /// A problem report.
+        report: ProblemReport,
+    },
+    /// Unable to determine a Catalyst ID of the registration.
+    ///
+    /// This can happen if a previous transaction ID in the registration is incorrect.
+    UnknownCatalystId,
+    /// A "fatal" error occurred during validation.
+    ///
+    /// This means that the validation wasn't performed properly (usually because of a
+    /// database failure) and we cannot process the given registration. This error is
+    /// propagated on a higher level, so there will be another attempt to index that
+    /// block.
+    Fatal(anyhow::Error),
+}
+
+impl From<anyhow::Error> for RbacValidationError {
+    fn from(e: anyhow::Error) -> Self {
+        RbacValidationError::Fatal(e)
+    }
+}
+
+/// Represents the result yielded by `update_chain` or `start_new_chain` upon successful
+/// execution.
+pub struct RbacValidationSuccess {
+    /// A Catalyst ID of the chain this registration belongs to.
+    pub catalyst_id: CatalystId,
+    /// A list of stake addresses that were added to the chain.
+    pub stake_addresses: HashSet<StakeAddress>,
+    /// A list of role public keys used in this registration.
+    pub public_keys: HashSet<VerifyingKey>,
+    /// A list of updates to other chains containing Catalyst IDs and removed stake
+    /// addresses.
+    ///
+    /// A new RBAC registration can take ownership of stake addresses of other chains.
+    pub modified_chains: Vec<(CatalystId, HashSet<StakeAddress>)>,
+    /// An updated registration chain.
+    pub chain: RegistrationChain,
+}
+
+/// Attempts to update an existing RBAC registration chain
+/// with a new CIP-509 registration, validating address and key usage consistency.
+///
+/// # Returns
+/// - `Ok((new_chain, validation_result))` if the chain was successfully updated and
+///   validated.
+///
+/// # Errors
+/// - Returns [`RbacValidationError::UnknownCatalystId`] if no Catalyst chain is found for
+///   `previous_txn`.
+/// - Returns [`RbacValidationError::InvalidRegistration`] if address/key duplication or
+///   validation inconsistencies are detected.
+pub async fn update_chain<Provider>(
+    reg: Cip509,
+    previous_txn: TransactionId,
+    is_persistent: bool,
+    provider: &Provider,
+) -> RbacValidationResult
```

Comment:
Based on the previous comment this whole functionality should be a part of the `RegistrationChain::update` method.